enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/StoreCopyClient.java

```diff
@@ -69,70 +75,72 @@ long copyStoreFiles( CatchupAddressProvider catchupAddressProvider, StoreId expe
         long lastTransactionId = prepareStoreCopyResponse.lastTransactionId();
         PrimitiveLongIterator indexIds = prepareStoreCopyResponse.getIndexIds().iterator();
         while ( indexIds.hasNext() )
         {
             long indexId = indexIds.next();
-            TerminationCondition terminationCondition = terminationConditions.get();
-            boolean successful;
-            do
+            persistentCall( new GetIndexFilesRequest( expectedStoreId, indexId, lastTransactionId ), copyHandler, addressProvider, terminationConditions.get(),
+                    exponentialBackoffStrategy );
+        }
+    }
+
+    private void persistentCall( CatchUpRequest request, CatchUpResponseAdaptor<StoreCopyFinishedResponse> copyHandler, CatchupAddressProvider addressProvider,
+            TerminationCondition terminationCondition, ExponentialBackoffStrategy exponentialBackoffStrategy ) throws StoreCopyFailedException
+    {
+        boolean successful;
+        do
+        {
+            try
             {
-                try
-                {
-                    AdvertisedSocketAddress from = addressProvider.primary();
-                    log.info( String.format( "Downloading snapshot of index '%s' from '%s'", indexId, from ) );
-                    StoreCopyFinishedResponse response =
-                            catchUpClient.makeBlockingRequest( from, new GetIndexFilesRequest( expectedStoreId, indexId, lastTransactionId ),
-                                    copyHandler );
-                    successful = successfulFileDownload( response );
-                }
-                catch ( CatchUpClientException | CatchupAddressResolutionException e )
-                {
-                    successful = false;
-                }
-                if ( !successful )
-                {
-                    log.error( "Failed to download files from index '%s'", indexId );
-                    terminationCondition.assertContinue();
-                }
+                AdvertisedSocketAddress from = addressProvider.secondary();
+                log.info( format( "Sending request '%s' to '%s'", request, from ) );
+                StoreCopyFinishedResponse response = catchUpClient.makeBlockingRequest( from, request, copyHandler );
+                successful = successfulFileDownload( response );
+            }
+            catch ( CatchUpClientException | CatchupAddressResolutionException e )
+            {
+                successful = false;
+            }
+            if ( !successful )
+            {
+                log.error( format( "Request failed '%s'", request ) );
+                terminationCondition.assertContinue();
+            }
+            else
+            {
+                log.info( format( "Request was successful '%s'", request ) );
             }
-            while ( !successful );
+            awaitTimeout( exponentialBackoffStrategy.newTimeout() );
```

Comment:
This is not used in the right way. You have to create the timeout using newTimeout(), whose lifecycle then spans the entirety of the operation (including retries), and then you use Timeout#getMillis() for every invocation of sleep, and after that Timeout#increment().

Reply:
I also put it in the wrong place. I'll fix this. It also hit me that I can be much smarter here depending on the response from the server. I think we only need to await on exception since a failed response from the server should mean that i should try another one.